### ISSUE TITLE
The performance of supervised learning differs significantly from unsupervised learning

### ISSUE THREAD

**Issue by HarrywillDr** [MaartenGr/BERTopic#620]
Hi,

Sorry to bother you.

I tried to train using 4260 with  207 labels (20:1)and tested on 110 test data,should be 5labels (20:1). But it doesn't perform well. Unsup methods, however, make sense.

Now, I'm trying to using a 10k training ser train,but data:label are (200:1), is that help?

How could I improve the performance?
Thanks!

**Comment by MaartenGr** [comment]
> Sorry to bother you.

No problem! Feel free to ask any questions. 

> I tried to train using 4260 with 207 labels (20:1)and tested on 110 test data,should be 5labels (20:1). 

I have a bit of trouble following you. Just to be sure, you trained a BERTopic instance with 4260 documents that all have positive labels (no -1 labels) and you tested with the `transform` function on 11 documents, correct? 

> But it doesn't perform well. 

What do you mean by that? Is there something specific that does not work well? 

Also, to get a bit more context, could you share your code here for training and testing your model?

**Comment by HarrywillDr** [comment]
Thanks for the reply! Sorry I have so many question!

> I have a bit of trouble following you. Just to be sure, you trained a BERTopic instance with 4260 documents that all have positive labels (no -1 labels) and you tested with the transform function on 11 documents, correct?

Yes, it is! The poor performance is unsupervised after parameter tuning, which can achieve very good clustering results. But supervised methods split most of the data into ‘-1’. 

Specifically, there are two results.

1. Clustering 90% of the data into one cluster and the rest into another or ‘-1’
2. Only a few of the data in 110 (maybe 7-8) have a cluster, but all the others are ‘-1’

Is this because there is a problem with my parameter settings？
For training, my 4000 documents belong to 5 classes 
For test, 110 data also belong to 5 classes, but the difference is that there are a small number of annoyment points in the test set.

When I pre-processed the data, I removed the words that relate to the topic from the sentence as stop words, will this have an effect on the results?

Here is the code.

```
from umap import UMAP
from hdbscan import HDBSCAN
from sklearn.feature_extraction.text import CountVectorizer

umap_model = UMAP(n_neighbors=20, n_components=5, min_dist=0.0, metric='cosine', random_state=42) # UMAP is an technique for dimensionality reduction

hdbscan_model = HDBSCAN(min_cluster_size=20, min_samples=5, metric='euclidean', cluster_selection_method='eom', prediction_data=True) # HDBSCAN used to cluster our documents into clusters of similar documents

vectorizer_model = CountVectorizer(ngram_range=(1, 1), stop_words="english") 
```

`from sentence_transformers import SentenceTransformer

# Document embedding model
sentence_model = SentenceTransformer("all-mpnet-base-v2", device="cuda")
embeddings = sentence_model.encode(docs,  show_progress_bar= True)`

`topic_model = BERTopic(embedding_model=sentence_model, umap_model=umap_model, hdbscan_model=hdbscan_model, vectorizer_model=vectorizer_model, 
                       calculate_probabilities= True, low_memory= False,n_gram_range = (1, 1), nr_topics="5", verbose = True)

topics, probs = topic_model.fit_transform(docs, embeddings, y=targets)`

`topics_reduced, probs_reduced = topic_model.reduce_topics(docs, topics, nr_topics=4)
topic_model.save(data_path + "trained_model")
loaded_model = BERTopic.load(data_path + "trained_model")`

`sentence_model = SentenceTransformer("all-mpnet-base-v2", device="cuda")
test_embeddings = sentence_model.encode(test_docs, show_progress_bar= True)
topics_test, probs_test = loaded_model.transform(test_docs, test_embeddings)`

Thank you so much!

**Comment by MaartenGr** [comment]
There are a couple of things here happening which could explain your performance. First, the following piece of code: 

> `topic_model = BERTopic(embedding_model=sentence_model, umap_model=umap_model, hdbscan_model=hdbscan_model, vectorizer_model=vectorizer_model, 
                       calculate_probabilities= True, low_memory= False,n_gram_range = (1, 1), nr_topics="5", verbose = True)

By setting `nr_topics=5` you are forcing the number of topics that are created to a very low value. Moreover, many of the topics that you would otherwise get are pushed into the `-1` topic, so any predictions later on are likely to fall into that outlier category. 

> topics_reduced, probs_reduced = topic_model.reduce_topics(docs, topics, nr_topics=4)

I am not entirely sure why this step is necessary if you already reduced the number of topics to 5 in the previous step. I would remove this step as further reducing the number of topics will only hurt performance. 

> Is this because there is a problem with my parameter settings？

Instead of focusing on the `nr_topics`, I would advise playing around with the `min_cluster_size` parameter in HDBSCAN as it directly impacts the number of topics that are created. 

> When I pre-processed the data, I removed the words that relate to the topic from the sentence as stop words, will this have an effect on the results?

Yes, this can negatively impact the resulting training process as these kind of transformers will need as much context as you can give it. Removing stop words is likely to make the embeddings less accurate. 

As a final note, you can use the tricks [here](https://maartengr.github.io/BERTopic/faq.html#how-do-i-reduce-topic-outliers) to reduce outliers and/or assign outliers to topics when testing.

**Comment by HarrywillDr** [comment]
Thank you so much!!!
